### PR TITLE
[Spark] Add GenerateIdentityValues UDF

### PR DESCRIPTION
**Pull request by c27kwan**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

This PR is part of https://github.com/delta-io/delta/issues/1959

In this PR, we introduce the `GenerateIdentityValues` UDF used for populating Identity Column values. The UDF is not used in Delta in this PR yet.

`GenerateIdentityValues` is a simple non-deterministic UDF which keeps a counter with the user specified `start` and `step`. It counts in increments of `numPartitions` so that it can be parallelized in different tasks.

## How was this patch tested?
New test suite and unit tests for the UDF.

## Does this PR introduce _any_ user-facing changes?
No.

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/IdentityColumn.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (c27kwan) [inline spark/src/main/scala/org/apache/spark/sql/delta/IdentityColumn.scala:1]: This file is part of the Scala API PR right now: https://github.com/delta-io/delta/pull/2857

Comment (c27kwan) [reply spark/src/main/scala/org/apache/spark/sql/delta/IdentityColumn.scala]: Merged in https://github.com/delta-io/delta/pull/2916 to reduce conflict between PRs.

File: spark/src/main/scala/org/apache/spark/sql/delta/GenerateIdentityValues.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, CodeGenerator, ExprCode, FalseLiteral}
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.types.{DataType, LongType}
+
+/**
+ * Returns the next generated IDENTITY column value based on the underlying
+ * [[PartitionIdentityValueGenerator]].
+ */
+case class GenerateIdentityValues(
+    generator: PartitionIdentityValueGenerator) extends LeafExpression with Nondeterministic {
```
Comment (fred-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/GenerateIdentityValues.scala:35]: Could we put the constructor on the same line and the extends on a separate line? I think it's more consistent with the rest of the codebase. :)

File: spark/src/main/scala/org/apache/spark/sql/delta/GenerateIdentityValues.scala [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    } else if (this.partitionIndex != partitionIndex) {
+      throw SparkException.internalError("Same PartitionIdentityValueGenerator object " +
+        s"initialized with two different partitionIndex [oldValue: ${this.partitionIndex}, " +
+        s"newValue: $partitionIndex]")
+
+    }
+  }
+
+  private def assertInitialized(): Unit = {
+    if (partitionIndex == -1) {
```
Comment (fred-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/GenerateIdentityValues.scala:131]: Could we put this on the same line as the method declaration? It'd reduce one level of indentation

File: spark/src/test/scala/org/apache/spark/sql/delta/GenerateIdentityValuesSuite.scala [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    val highWaterMarks = Seq(None, Some((start + 100 * step).toLong))
+    val df = spark.range(1, size + 1, 1, slice).toDF(colName)
+    highWaterMarks.foreach { highWaterMark =>
+      // Create two GenerateIdentityValues expressions that share the same state. They should
+      // generate distinct values.
+      val gev = GenerateIdentityValues(start, step, highWaterMark)
+      val gev2 = gev.copy()
+      verifyIdentityValues(
+        df.select(new Column(
+            If(GreaterThan(col(colName).expr, Literal(10)), gev, gev2)).alias(colName)),
```
Comment (fred-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/GenerateIdentityValuesSuite.scala:97]: Could we make the Literal a named argument?

**Review by fred-db — APPROVED** [review]
LGTM, just some nits :)

**Pull request merged**